### DISCUSSION TITLE
Make Basic HTTP Auth for endpoint conditional

### DISCUSSION THREAD

**Discussion by Koliham** [fastapi/fastapi#6958] [Questions] [answered]
I have a configuration file and depending on a flag set in this config, I want to make the endpoints have a Basic HTTP Auth as a dependency or to be called without any Auth parameters.
https://fastapi.tiangolo.com/advanced/security/http-basic-auth/

Using things like:
```
if config["auth_needed"]==False:
   return "serviceaccount"
else:
   # do some credentials check
```

in the function that checks the credentials allows any given username/password as valid, but still requires Basic Auth information, to be filled.
I tried with decorators, but couldn't enrich the function argument with an additional credentials: `HTTPBasicCredentials = Depends(security)`

**Comment by includeamin** [comment]
please follow the [Issue instruction](https://github.com/tiangolo/fastapi/issues/new/choose).
we need more information to help you.
thanks.

**Comment by Koliham** [comment, marked as answer]
Ok, I solved it:
The classical approach for having Basic HTTP Auth like on the FastAPI page is:

```
@app.get("/users/me")
def get_current_username (username: str = Depends(get_current_username)):
    return {"username": username}
```

for the endpoint and for the checking function:
```
def get_current_username(credentials: HTTPBasicCredentials = Depends(security)):
    correct_username = secrets.compare_digest(credentials.username, "stanleyjobson")
   [...]
```

What I did was to create a new security function `optional_security`:

```
async def optional_security(request: Request):
    config["auth_needed"]==True:
        return await security(request)
    else:
        return None
```

Now the get_current_username function for the secured route has to be fixed:

```
def get_current_username(credentials: Optional[HTTPBasicCredentials] = Depends(optional_security)):
    if config["auth_needed"]==False:
        return "whatever"
    # do the actual credentials check
   [...]
```

Now I can switch the Basic Authentication on and off. for the endpoints with a config file.
The caveat is that OpenAPI does not show the lock symbol for endpoints, which need authentication, but it still asks for it, when you execute it, so its fine.

**Comment by Kludex** [comment]
@Koliham If you've solved it, do you mind closing it? :kissing:

**Comment by tiangolo** [comment]
Thanks for the help here @Kludex and @includeamin ! 👏  🙇 

Thanks for reporting back and closing the issue @Koliham  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
